File: android/app-v2/src/main/java/com/smarthome/livingroom_v2/command/dispatcher/TaskDispatcher.kt
package com.smarthome.livingroom_v2.command.dispatcher

import android.util.Log
import com.smarthome.livingroom_v2.command.Task
import com.smarthome.livingroom_v2.command.runtime.EdgeRuntime
import com.smarthome.livingroom_v2.command.runtime.EdgeRuntimeNode
import com.smarthome.livingroom_v2.command.runtime.TaskExecutionResult

interface TaskDispatcher {
    suspend fun dispatch(task: Task, node: EdgeRuntimeNode): TaskExecutionResult
}

class LocalTaskDispatcher(
    private val runtime: EdgeRuntime,
) : TaskDispatcher {
    override suspend fun dispatch(task: Task, node: EdgeRuntimeNode): TaskExecutionResult {
        Log.i(TAG, "dispatch taskId=${task.taskId} → node=${node.nodeId}")
        return runtime.execute(task, node)
    }

    companion object {
        private const val TAG = "TaskDispatcher"
    }
}

File: ios/LivingRoomEdge/LivingRoomEdge/Command/Dispatcher/TaskDispatcher.swift
import Foundation

@MainActor
protocol TaskDispatcher: AnyObject {
    func dispatch(task: EdgeTask, node: EdgeRuntimeNode) async -> TaskExecutionResult
}

@MainActor
final class LocalTaskDispatcher: TaskDispatcher {
    private let runtime: EdgeRuntime

    init(runtime: EdgeRuntime) {
        self.runtime = runtime
    }

    func dispatch(task: EdgeTask, node: EdgeRuntimeNode) async -> TaskExecutionResult {
        NSLog("[TaskDispatcher] dispatch taskId=%@ → %@", task.taskId, node.nodeId)
        return await runtime.execute(task: task, node: node)
    }
}
